src/main/scala/com/fulcrumgenomics/bam/ErrorRateByReadPosition.scala

```diff
@@ -0,0 +1,280 @@
+  }
+
+  protected def maybeReverseComplement(bases: Array[Byte], rc: Boolean): Array[Byte] = {
+    if (rc) SequenceUtil.reverseComplement(bases)
+    bases
+  }
+
+  protected def maybeReverse(qualities: Array[Byte], rc: Boolean): Array[Byte] = {
+    if (rc) SequenceUtil.reverse(qualities, 0, qualities.length)
+    qualities
+  }
+}
+
+/** Computes the error rate using a reference FASTA.  Does not rely on the MD tag. */
+private class ErrorRateByReadPositionWithReferenceCounter
+( ref: PathToFasta,
+  includeInsertions: Boolean,
+  minBaseQuality: Int = 0
+) extends ErrorRateByReadPositionCounter(includeInsertions=includeInsertions, minBaseQuality=minBaseQuality) {
+  import scala.collection.JavaConversions.iterableAsScalaIterable
+
+  val referenceFile     = ReferenceSequenceFileFactory.getReferenceSequenceFile(ref)
+  var referenceSequence = referenceFile.getSequence(referenceFile.getSequenceDictionary.getSequence(0).getSequenceName)
+
+  protected def updateErrors(rec: SAMRecord, readIndex: Int): Unit = {
+    // Get the reference bases
+    val referenceBases = referenceSequence.getBases
+
+    // We proceed through the read in the order it was aligned, but if the read is on the reverse strand, we will
+    // need to invert the read position since the last base in the read will be the first base sequenced
+    val readOffsetIncrement = if (rec.getReadNegativeStrandFlag) -1 else 1
+    val readLength = rec.getReadLength
+
+    // Get the read bases and qualities.
+    val read  = rec.getReadBases
+    val quals = rec.getBaseQualities
+
+    // For each alignment block
+    rec.getAlignmentBlocks.foreach { block =>
+      var refPosition  = block.getReferenceStart - 1
+      var readPosition = block.getReadStart - 1
+      val readEnd      = readPosition + block.getLength
+      var readOffset   = if (rec.getReadNegativeStrandFlag) readLength - readPosition - 1 else readPosition
+
+      // Matches/mismatches
+      while (readPosition < readEnd) {
+        val readBase = read(readPosition)
+        val qual     = quals(readPosition)
+        val refBase  = referenceBases(refPosition)
+        incrementIfValid(readIndex=readIndex, readOffset=readOffset, readBase=readBase, qual=qual, refBase=refBase)
+        refPosition += 1
+        readPosition   += 1
+        readOffset  += readOffsetIncrement
+      }
+    }
+
+    // Handle insertions
+    var readOffset = if (rec.getReadNegativeStrandFlag) rec.getReadLength - 1 else 0
+    rec.getCigar.getCigarElements.foreach { elem =>
+      val op: CigarOperator = elem.getOperator
+      if (op == CigarOperator.INSERTION) {
+        var i = 0
+        while (i < elem.getLength) {
+          val qual = quals(readOffset)
+          increment(readIndex = readIndex, readOffset = readOffset, qual=qual, isError = includeInsertions)
+          readOffset += readOffsetIncrement
+          i += 1
+        }
+      }
+      else {
+        readOffset += readOffsetIncrement * elem.getLength
```

Comment:
This probably wants to have `else if (op.consumeReadBases)`.  Otherwise, e.g. padding and deletions will do strange things.